pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go

```diff
@@ -344,11 +361,21 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 
 	for _, rprt := range rprts {
 		if rprt != nil {
-			c.Logger.Infof("Creating a new TaskRun object %s", rprt.TaskRunName)
-			rprt.TaskRun, err = c.createTaskRun(c.Logger, rprt, pr, as.StorageBasePath(pr))
-			if err != nil {
-				c.Recorder.Eventf(pr, corev1.EventTypeWarning, "TaskRunCreationFailed", "Failed to create TaskRun %q: %v", rprt.TaskRunName, err)
-				return xerrors.Errorf("error creating TaskRun called %s for PipelineTask %s from PipelineRun %s: %w", rprt.TaskRunName, rprt.PipelineTask.Name, pr.Name, err)
+			if rprt.ResolvedConditionChecks == nil || rprt.ResolvedConditionChecks.IsSuccess() {
+				c.Logger.Infof("Creating a new TaskRun object %s", rprt.TaskRunName)
+				rprt.TaskRun, err = c.createTaskRun(c.Logger, rprt, pr, as.StorageBasePath(pr))
+				if err != nil {
+					c.Recorder.Eventf(pr, corev1.EventTypeWarning, "TaskRunCreationFailed", "Failed to create TaskRun %q: %v", rprt.TaskRunName, err)
+					return xerrors.Errorf("error creating TaskRun called %s for PipelineTask %s from PipelineRun %s: %w", rprt.TaskRunName, rprt.PipelineTask.Name, pr.Name, err)
+				}
+			} else if !rprt.ResolvedConditionChecks.HasStarted() {
+				for _, rcc := range rprt.ResolvedConditionChecks {
+					rcc.ConditionCheck, err = c.makeConditionCheckContainer(c.Logger, rprt, rcc, pr)
+					if err != nil {
+						c.Recorder.Eventf(pr, corev1.EventTypeWarning, "ConditionCheckCreationFailed", "Failed to create TaskRun %q: %v", rcc.ConditionCheckName, err)
+						return xerrors.Errorf("error creating ConditionCheck container called %s for PipelineTask %s from PipelineRun %s: %w", rcc.ConditionCheckName, rprt.PipelineTask.Name, pr.Name, err)
+					}
+				}
```

Comment:
what do you think, could this chunk of code be a separate function? (outside the reconciler)

Reply:
Simplified this slightly (reduced one level of indentation). This block requires a bunch of the reconciler field (logger, recoder, the clientset). In the create Taskrun function, we use the clientset for both creating as well as updating, so it'll be hard to get it out of the reconciler. I'll play around with this some more, but what do you think about refactoring this in a follow-up?

Reply:
>  I'll play around with this some more, but what do you think about refactoring this in a follow-up?

Sounds good to me :D Thanks for looking into it :)